### DISCUSSION TITLE
Expecting token of type 'XX' but found `XX'

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1424] [Q&A] [answered]
![image](https://github.com/eclipse-langium/langium/assets/26995040/fc7ea43d-b848-48c6-b623-f0a886feb203)

**Comment by msujew** [comment, marked as answer]
@zrwsmd the token `ID` shadows `Identifier`, since it matches almost the same content, but comes before it in the terminal listing. Terminals are matched on a first-fit basis. Generally, different terminals should not intersect.

**Reply by zrwsmd** [reply]
so if i need to match both  `INT`  `qw_3`  `qw3`   `aa`  this four character string,how should i define a Terminals ,just can not write it correctly

grammar StructuredText

entry Document:
    (stStatements+=Var);


Var:
 'VAR' name=Identifier ':' type=Identifier ';'  'END_VAR'
 ;


// identifier characters
terminal fragment Underscore: /_/;
terminal fragment Letters: /[a-zA-Z]+/;
terminal fragment Digits: /[0-9]+/;
terminal Identifier: ((Underscore)? ((Digits) | Letters));



// hidden
hidden terminal WS: /\s+/;
hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal ML_COMMENT_ST: /\(\*[\s\S]*?\*\)/;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;

**Reply by msujew** [reply]
You can just use a simple regex terminal for that like `terminal ID: /[a-zA-Z][\w]*/`.
